### ISSUE TITLE
Proposal: @JSExportStatic: static fields and methods for JS classes

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#2702]
This is a proposal to address #1902. Besides, it also encompasses the definition of static methods on Scala.js-defined JS classes, as well as exporting top-level vals and vars. In fact, it generalizes `@JSExportTopLevel` in two directions: a) as `@JSExportStatic` to export as static member rather than top-level member and b) to apply to val/vars as well as methods.

## `@JSExportStatic` for methods

The annotation `@JSExportStatic` can be used on a `def` of a Scala top-level `object` (similarly to `@JSExportTopLevel` as in 0.6.14) but only if the companion class of that object is a Scala.js-defined JS class. For example:
```scala
@ScalaJSDefined
class Foo extends js.Object

object Foo { // a Scala object, not a JS object
  @JSExportStatic
  def bar(x: Int): Int = x + 1
}
```
The semantics of such an export is that a static method `"bar"` will be defined on the JS class `Foo`, which is a typical export-like forwarder to `Foo.bar`. Overloading is handled as usual for exports. With those definitions, the following snippet would work:
```scala
val fooClass: js.Dynamic = js.constructorOf[Foo]
val y = fooClass.bar(5)
println(y) // 6

fooClass.bar = (x: Int) => x * 2 // patch it, because we can!
println(fooClass.bar(5)) // 10
println(Foo.bar(5)) // still 6! We only patched the forwarder, not the Scala def
```
So far, just like `@JSExportTopLevel`, only defs with parentheses can be exported as static.

## Exporting getters and setters as static

Unlike `@JSExportTopLevel`, we can easily generalize `@JSExportStatic` to support getters (`def` without parentheses) and setters (`def` with a name ending with `_=`), as we can define static properties on the companion class.

This is an extension that does not necessarily need to be supported in a first implementation of the proposal.

## Exporting static fields (vals and vars)

`@JSExportStatic` can also be used on `val`s and `var`s in top-level Scala objects whose companion class is a Scala.js-defined JS class. In that case, an actual *field* is created on the JS class (*not* a getter/setter as is the case for `@JSExport`). The synthetic Scala getters and setters for that field are rewritten to access the static field (and box/unbox it as needed).

This can be used as follows:
```scala
@ScalaJSDefined
class Foo extends js.Object

object Foo { // a Scala object, not a JS object
  @JSExportStatic
  val alpha: Int = 1
  @JSExportStatic
  var beta: String = "hello"
}

val fooClass: js.Dynamic = js.constructorOf[Foo]
println(fooClass.alpha) // 1
println(fooClass.beta) // hello
fooClass.beta = "beta"
println(Foo.beta) // beta -- note that this accesses through the Scala object
fooClass.beta = 5 // wrong type
println(Foo.beta) // UndefinedBehaviorError inside the getter Foo.beta
```
There is an important difference between the `def` case and the `val`/`var` case. In the `def` case, the JS static method is a forwarder to the Scala method. In the `val`/`var` case, it is the other way around: the Scala accessors are forwarders to the JS field. This caveat is discussed later. For a "normal" usage of those things, the difference should not be noticeable.

## `@JSExportTopLevel` for vals and vars

There is an undeniable similarity between `@JSExportTopLevel` and `@JSExportStatic` for defs. They are basically equivalent in all aspects; a static export is simply nested in a namespace that happens to be a Scala.js-defined JS class. It seems logical that we should be able to "backport" the val/var case of `@JSExportStatic` to `@JSExportTopLevel`. And indeed, we can.

`@JSExportTopLevel` can be used on a val/var inside a top-level (static) Scala object, as follows:
```scala
object Foo {
  @JSExportTopLevel("alpha")
  val alpha: Int = 1
  @JSExportTopLevel("beta")
  var beta: String = "hello"
}
```
Just like `@JSExportStatic` causes the Scala accessors to redirect to JS static fields, `@JSExportTopLevel` does the same but with "top-level" JS fields. In fact, they are only top-level as far as the ES module implicitly defined by a Scala.js codebase is concerned. Said field is then exported from the ES module. The above therefore has the semantics of the following ES code (we omit *initializers* for now)
```javascript
let $t_alpha = 0; // the zero of the type, not the initial value
let $t_beta = null;

export { $t_alpha as alpha, $t_beta as beta };
```
and the accessors for `Foo.alpha` and `Foo.beta` access `$t_alpha` and `$t_beta`.

Note that `alpha` is an exported *let* and not a *const*. We need it to be a *let* because it has to be initialized by the constructor of `Foo$`. But that's not really a big deal, because from the outside of an ES module, an imported *let* is not mutable; it is read-only. This is in fact exactly the semantics we want.

## `@JSExportTopLevel` and getters/setters

Since an ES module *cannot* export a top-level property (getter/setter), `@JSExportTopLevel` is not applicable to `def`s without parentheses nor `def`s with a name ending in `_=`.

## The problem of initializers

So far, I have ignored the problem of initializing exported top-level and static fields. In fact, it is not terribly difficult to address. First, all fields are of course created with the zero of their type. They are truly *initialized* by the constructor of the `object Foo`, when it runs.

Of course, we want it to run "soon enough". For an `@JSExportStatic`, we make sure it runs soon enough by saying that:

If a Scala.js-defined JS class has at least one static field (coming from its companion object), then after the JS class value has been created, but before it is returned from the class value accessor (see #2685), the module accessor of its companion object is executed. This ensures that the constructor of the latter is executed, which will properly initialize the field.

For an `@JSExportTopLevel`, a call to the module accessor is automatically inserted as top-level statements in the output ES module (or CommonJS module, etc.). If there are several objects with `@JSExportTopLevel` fields, the order in which they are executed is left to the implementation. It can be non-deterministic.

## The "reverse" aspect of fields versus methods

Recall that for static and top-level defs, the JS method is a forwarder to the Scala method; whereas for static and top-level vals and vars, the Scala accessors are forwarders to the JS field. This is potentially counter-intuitive, but is, I think, the best compromise.

We have to balance it with the two seemingly more reasonable alternative: either in all cases JS is a forwarder to Scala; or in all cases Scala is a forwarder to JS.

The former (JS is a forwarder to Scala) is simply not an option for fields. Indeed, a JS field simply *cannot* be a forwarder (by definition). One would argue that it could be a pair of getter/setter (after all that's what we do for `@JSExport`ed vals and vars) but this does not fly: for static fields, it means we do not have any interop feature whose semantics is to define an actual static *field*. And for top-level fields, that's even less of an option, because ES exports cannot be getter/setters.

We are therefore left with the alternative that for `def`s, Scala methods would be a forwarder to the JS methods. This is not completely nonsensical (unlike JS fields being forwarders), but it poses a major problem for overloading. Indeed, the compile-time overloading semantics of defs in a Scala object have to be forfeited if Scala defs become forwarders to the JS defs. This is really bad, especially if one or more of those methods actually *overrides* something from a (Scala) superclass. Then the compile-time overloading semantics of the superclass are broken by the subclass, which is not acceptable.

Fields do not suffer from the overloading problem, which means that this is not an issue for fields. They are not completely exempt of semantics changes, though. Indeed, there is now a possibility for some JS code to store a value of an illegal type in the field (as demonstrated above). This can cause an `UndefinedBehaviorError` upon reading the field from the Scala side, which would never happen in pure Scala code. However, this seems fairly minor compared to the overloading problem.

**Comment by sjrd** [comment]
Comments welcome!
/cc @gzm0

**Comment by sjrd** [comment]
When emitting in ES6 output mode, yes, because a class value's prototype is its parent class value in ECMAScript 2015.

When emitting in ES5.1 output mode, no. At least this proposal does not mention anything in favor or against it. We might have to adjust our ES5.1 output so that it sets the prototype of class values, but it is actually quite orthogonal to this proposal.

**Comment by chandu0101** [comment]
>When emitting in ES5.1 output mode, no. At least this proposal does not mention anything in favor or against it. We might have to adjust our ES5.1 output so that it sets the prototype of class values, but it is actually quite orthogonal to this proposal.

ok , its nice to have that in ES 5.1 output mode in future, for now  hello Babel! ...

**Comment by sjrd** [comment]
Yes, I went through the same reasoning which led me to rule out Scala.js-defined JS companion objects. I think I intended to write that in another Discussion point in the proposal, but obviously I have forgotten ^^ But you precisely laid out the reasons :)

**Comment by sjrd** [comment]
This proposal was fully implemented by a combination of #2705, #2716 and #2731.
